packages/jupyter-ai/jupyter_ai/callback_handlers/metadata.py

```diff
@@ -23,4 +34,15 @@ def on_llm_end(self, response: LLMResult, **kwargs) -> None:
         if not (len(response.generations) and len(response.generations[0])):
             return
 
-        self.jai_metadata = response.generations[0][0].generation_info or {}
+        metadata = response.generations[0][0].generation_info or {}
+
+        # Convert any non-serializable objects in metadata
+        serializable_metadata = {}
+        for key, value in metadata.items():
+            try:
+                json.dumps(value)
+                serializable_metadata[key] = value
+            except (TypeError, ValueError):
+                serializable_metadata[key] = convert_to_serializable(value)
+
+        self.jai_metadata = serializable_metadata
```

Comment:
One alternative would be something like:


```suggestion
        self.jai_metadata = json.loads(
            json.dumps(
                metadata,
                default=lambda o: f"<non-serializable: {type(o).__qualname__}>"
            )
        )
```

Reply:
Thanks @JanusChoi and @krassowski for offering fixes to this issue. I have tested both fix versions and both work. The main difference as I see it is that the solution from @JanusChoi populates a `Dict` (which is serializable) with the values being passed to the `Message` class (which is not serializable). The `message.dict()` object is as follows:
<img width="909" alt="image" src="https://github.com/user-attachments/assets/8b3e1c58-6242-40e6-beda-de5f34a61cd3">
We can see the contents of the `message` attribute at the end of the item above. (I explicitly added a print statement to show the `message.dict()` object, so it is not logged otherwise. 

The modification offered by @krassowski specifically marks the `message` object as non-serializable. The `message.dict()` looks like the following: 
<img width="912" alt="image" src="https://github.com/user-attachments/assets/3b42970e-fae4-402e-bf3d-fb5feeedf746">
Again, this is not logged unless additional statements are added. 

So the question is, which approach is best? As @krassowski notes, this is not specific to Ollama, but we note that it is not an issue with other model providers. Moreover, this issue arose more recently, so it may be useful to chase down what changed to trigger this.

Reply:
To spell it out, it's a trade-off between less maintenance cost in my approach but less detailed info in metadata vs more detailed info but if any other provider starts returning non-serializable stuff that does not have `.dict` method or `__dict__` property then it breaks again.

Do I get it right that the metadata is only used by some jupyter-ai operators for telemetry and by default when users install it then it is a non-feature?

Reply:
I guess we could combine both approaches by adjusting the `default` to first check if there is `.dict` or `__dict__` and only then returning `<non-serializable>`

Reply:
@krassowski Corrct, this _is_ coming from a telemetry add on. Let's see what @dlq thinks as well.

Reply:
Thank you both for iterating on this thread discussion. I agree that it makes sense to combine both approaches. I'll close this thread out and start a new one that summarizes your conclusions.